### ISSUE TITLE
consider use `json.RawMessage` for ext fields

### ISSUE THREAD

**Issue by csyangchen** [mxmCherry/openrtb#8]
for now, ext fields are of type `interface{}`, which I see two issues in production code:

- bad in decoding performance, prevent effective use of [ffjson](https://github.com/pquerna/ffjson), see <https://github.com/pquerna/ffjson/issues/125>
- read value from `interface{}` is not easy, sometimes we have to do conversion from `interface{}` to custom ext struct type

consider use `ext *json.RawMesage` fields to defer decoding to later stage, *if* there actually is a need to peek into `ext` fields.

**Comment by mxmCherry** [comment]
Already thought about that - `interface{}` is convenient only at SSP side (for encoding). But I don't think SSPs tend to use abstract/unmodified libs for this.

As far as I know, most exchanges add `ext` only on `BidRequest` type (btw, you got `ext` not on `BidRequest` message?), so originally I planned approach like:

```
ext := new(SpecificExchangeExt) // ext var is just to avoid type checking
req := &openrtb.BidRequest{Ext: ext}

_ = json.Unmarshal([]byte(`{
"ext": {
	"foo": 42
}
}`), req)

fmt.Printf("%+v\n", ext)
```

Another approach is to fork this repo for every exchange (so you may use specific `ext` types right in OpenRTB structs).

But I agree - `json.RawMessage` would be better than `interface{}` in a "base" (this) package (as it's mostly intended for DSPs).

I'll think over it on this weekend (Sat-Sun), probably will add/use custom type (like [here](https://github.com/bsm/openrtb/pull/26/files#diff-e5f928119404365bf9bc159c1b2bc9f5R5)), as using pointers with `[]byte` is usually not very convenient/beautiful.

**Comment by mxmCherry** [comment]
Implemented, released as `v3.0.0`: https://github.com/mxmCherry/openrtb/releases/tag/v3.0.0
